### ISSUE TITLE
Java lambda function classpath mix up with localstack-utils-fat.jar

### ISSUE THREAD

**Issue by albuch** [localstack/localstack#1160]
I'm writing a Lambda function with Java8 and try to use AWS SDK for Java to access a S3 bucket. 
The AWS Java SDK for S3 is bundled with the java function in version 1.11.510. The function does work on AWS.

When executing the function on localstack the function fails with the following error when trying to access the S3 resource:
```
Exception: Lambda process returned error status code: 1. Output:
Exception in thread "main" java.lang.NoSuchFieldError: SERVICE_ID
        at com.amazonaws.services.s3.AmazonS3Client.createRequest(AmazonS3Client.java:4655)
        at com.amazonaws.services.s3.AmazonS3Client.createRequest(AmazonS3Client.java:4630)
        at com.amazonaws.services.s3.AmazonS3Client.listBuckets(AmazonS3Client.java:976)
        at com.amazonaws.services.s3.AmazonS3Client.listBuckets(AmazonS3Client.java:984)
        at com.App.handleRequest(App.java:22)
        at cloud.localstack.LambdaExecutor.main(LambdaExecutor.java:100)
```
When checking the class that defines the field `SERVICE_ID` with `HandlerContextKey.class.getProtectionDomain().getCodeSource()` the output is 
`file:/home/user/.local/lib/python2.7/site-packages/localstack/infra/localstack-utils-fat.jar` which is not from the bundled jar file of the Lambda function but from localstacks own jar files.
This results in incompatibilities between different class versions.

Is there a requirement that localstack only works with a specific AWS Java SDK version? If not how can I get the function to work on localstack?

I'm using localstack v0.8.10 on Windows 10 via Windows Subsystem for Linux (WSL) Ubuntu 18.04.1 LTS.

possible duplicate of #952

UPDATE:
After changing the lambda function to use aws-java-sdk v1.11.310 which is the version included in localstack-utils-fat.jar the lambda function executes properly.

This is still an isolation issue that might result in issues with other classes where no easy workaround is available.

One would rather want to load the jar file in it's own classpath (hacky) or apply a proper modularization via OSGi or Java 9 modules (JIGSAW)

**Comment by albuch** [comment]
possible duplicate of #952

**Comment by albuch** [comment]
After changing the lambda function to use aws-java-sdk `v1.11.310` which is the version included in `localstack-utils-fat.jar` the lambda function executes properly.

This is still an isolation issue that might result in issues with other classes where no easy workaround is available.

**Comment by whummer** [comment]
Thanks for the detailed bug report @albuch - do you think you could help out with a pull request here?

**Comment by whummer** [comment]
Thanks for confirming the workaround @albuch . It seems to me that this issue could be solved by fixing the order of entries in the classpath. A push has been fixed in #1817 - can you please give it a try once the PR is merged? Also, if you could share a small self-contained example here that allows to reproduce the issue, that would be great. Thanks

**Comment by whummer** [comment]
Closing this issue - please re-open if the problem persists. Thanks

**Comment by d4vidB** [comment]
I think I am still encountering the same issue, my Java lambda function works fine on AWS but in LocalStack, I get the following error when building my DynamoDB client:

`java.lang.NoSuchMethodError: com.amazonaws.protocol.json.JsonErrorShapeMetadata.withExceptionUnmarshaller(Lcom/amazonaws/transform/JsonErrorUnmarshaller;)Lcom/amazonaws/protocol/json/JsonErrorShapeMetadata`

I am using AWS SDK 1.11.683.

**Comment by krrg** [comment]
Unfortunately, I'm having the same problem with the `localstack-utils-fat.jar` somehow taking priority over my own jar for class resolution, leading to incompatible versions of the AWS SDK.

A related issue seems to be: https://github.com/localstack/localstack/issues/1031
